### PR TITLE
Harness

### PR DESCRIPTION
**Pull request by pbx0**
Factor the test harness out into a separate package.

### CODE REVIEW

File: kola/harness.go [hunk 1]
```diff
@@ -20,53 +20,57 @@ import (
...
+	Platforms   []string // whitelist of platforms to run test against -- defaults to all
 }
 
 // runs and sets up environment for tests specified in etcdtests pkg
-func runEtcd(args []string) int {
-	if len(args) != 0 {
-		fmt.Fprintf(os.Stderr, "No args accepted\n")
+func RunTests(args []string) int {
+	if len(args) != 1 {
+		fmt.Fprintf(os.Stderr, "No test group specified. Use 'all' to run all tests\n")
```
Comment (marineam) [inline kola/harness.go]: Why not just make 'all' the default?

File: kola/harness.go [hunk 2]
```diff
@@ -20,53 +20,57 @@ import (
...
-
-This must run as root!
-`}
-
-func init() {
-	cli.Register(cmdEtcd)
+type Test struct {
+	Run         func(platform.Cluster) error
+	Name        string
+	Group       string // used to manually group related tests
```
Comment (marineam) [inline kola/harness.go]: Other test suites often use name pattern matching and/or some sort of tag scheme such as one indicating the execution time like 'long' vs 'short'. We could probably combine Group and Platforms into a generic "tags" scheme. If you want to simplify this PR you can just drop Group for now but add pattern matching on test names.

File: kola/tests/exampleTest [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+package example
+
+import "github.com/coreos/mantle/harness"
+
+// This is necessary somewhere in the package to register all tests you wish to make available
+// for kola to run.
+func init() {
+	harness.Tests = append(harness.Tests, []harness.Test{
+		Test{
+			Name:        "versionCheck",
```
Comment (marineam) [inline kola/tests/exampleTest]: Lets name 'Name' practical as a global name, useful to the end user. Something like "EtcdVersionCheck" would work.

File: kola/tests/exampleTest [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+			Group:    "version",
+			// This will group sets of tests together to be run by a
+			// kola subcommand. It'd be nice to do it automatically by
+			// packages, but I think that would require codegen.
+		},
+
+		// Multiple tests configuations can be specified for a single
+		// run function. Each trial gets a new cluster.
+		Test{
+			Name:        "versionCheck",
```
Comment (marineam) [inline kola/tests/exampleTest]: Names should be unique, otherwise what's the purpose of a name?

File: kola/tests/exampleTest [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+			Run:         versionCheck,
+			Discovery:   true,
+			ClusterSize: 5,
+			CloudConfig: `#cloud-config...`,
+			Group:    "version",
```
Comment (marineam) [inline kola/tests/exampleTest]: I don't see how a group name like this is useful. Just drop it please.

File: kola/etcdregistry.go [hunk 6]
```diff
@@ -12,37 +12,36 @@
...
-		Discovery:   true,
-		ClusterSize: 3,
-		Name:        "etcd2 discovery",
-		EtcdVersion: 2,
-		CloudConfig: `#cloud-config
+		// test etcd discovery with 2.0 with new cloud config
+		Test{
+			Run:         etcd.DiscoveryV2,
+			ClusterSize: 3,
+			Name:        "etcd2 discovery",
```
Comment (marineam) [inline kola/etcdregistry.go]: Don't use spaces in test names, that'll make it easier to specify them as a command line option.

File: kola/harness.go (hunk 2 above)
Comment (pbx0) [inline kola/harness.go]: Sounds good for now. Can pattern matching be as simple as `kola run name-substring` that matches any test whose name contains it? Or would something like `kola run regex-pattern`be better?

Comment (marineam) [inline kola/harness.go]: Lets do globbing, that is easier to do on command line than regex. So kola run 'Etcd*' to do etcd tests. Also lets make test names start with uppercase. Later if we add tags like 'long' or 'short' they can be lower case so kola run short isn't ambiguous.

**Comment by pbx0** [comment]
PTAL:
All tests run by default. Groups are gone and names are patterned matched for the time being. Test names should be unique and uppercase. Deleted example test for now. Examples should just go in documentation.

File: kola/flags.go [hunk 7]
```diff
@@ -0,0 +1,33 @@
+package kola
```
Comment (marineam) [inline kola/flags.go]: Missing license header

**Comment by marineam** [comment]
LGTM

**Pull request merged**